### PR TITLE
PERF: Rolling._apply

### LINKED ISSUE #42736: PERF: Rolling._apply iterates over blocks _and_ columns
@mroeschke looking at BaseWindow._apply we're spending a lot of time in apply_along_axis in our asvs (results posted below).  IIUC whats happening is that homogeneous_func is passed to BlockManager.apply thereby iterating over Blocks, and then within that call it iterates over columns.

My intuition is that we should either iterate over blocks or columns but not both.  Is there a reason to do both?

Not sure about UDFs, but for mean/sum/... it seems like we'd have to edit the cython functions to iterate over columns there, right?

```
from asv_bench.benchmarks.rolling import *

self = Engine()
self.setup("DataFrame", "float", sum, "cython", "mean")

%prun -s cumtime for n in range(1000): self.time_rolling_methods('DataFrame', 'float', sum, 'cython', 'mean')

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.325    0.325 {built-in method builtins.exec}
        1    0.001    0.001    0.325    0.325 <string>:1(<module>)
     1000    0.003    0.000    0.324    0.000 rolling.py:68(time_rolling_methods)
     1000    0.002    0.000    0.302    0.000 rolling.py:1802(mean)
     1000    0.002    0.000    0.299    0.000 rolling.py:1291(mean)
     1000    0.002    0.000    0.295    0.000 rolling.py:479(_apply)
     1000    0.007    0.000    0.291    0.000 rolling.py:408(_apply_blockwise)
     1000    0.003    0.000    0.268    0.000 managers.py:276(apply)
     1000    0.003    0.000    0.204    0.000 blocks.py:364(apply)
     1000    0.002    0.000    0.165    0.000 rolling.py:425(hfunc)
     1000    0.004    0.000    0.149    0.000 rolling.py:514(homogeneous_func)
14000/3000    0.013    0.000    0.140    0.000 {built-in method numpy.core._multiarray_umath.implement_array_function}
     1000    0.001    0.000    0.136    0.000 <__array_function__ internals>:2(apply_along_axis)
     1000    0.015    0.000    0.134    0.000 shape_base.py:267(apply_along_axis)
     1000    0.002    0.000    0.074    0.000 rolling.py:520(calc)
     1000    0.010    0.000    0.058    0.000 managers.py:539(_combine)
     1000    0.007    0.000    0.050    0.000 indexers.py:73(get_window_bounds)

# With ArrayManager so we only iterate over columns:
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.198    0.198 {built-in method builtins.exec}
        1    0.001    0.001    0.198    0.198 <string>:1(<module>)
     1000    0.003    0.000    0.197    0.000 rolling.py:68(time_rolling_methods)
     1000    0.002    0.000    0.174    0.000 rolling.py:1802(mean)
     1000    0.002    0.000    0.171    0.000 rolling.py:1291(mean)
     1000    0.002    0.000    0.167    0.000 rolling.py:479(_apply)
     1000    0.007    0.000    0.163    0.000 rolling.py:408(_apply_blockwise)
     1000    0.005    0.000    0.139    0.000 array_manager.py:194(apply)
     1000    0.002    0.000    0.100    0.000 rolling.py:425(hfunc)
     1000    0.004    0.000    0.086    0.000 rolling.py:514(homogeneous_func)
     1000    0.002    0.000    0.071    0.000 rolling.py:520(calc)
     1000    0.006    0.000    0.047    0.000 indexers.py:73(get_window_bounds)
     2000    0.001    0.000    0.039    0.000 <__array_function__ internals>:2(clip)
7000/3000    0.003    0.000    0.038    0.000 {built-in method numpy.core._multiarray_umath.implement_array_function}
     2000    0.002    0.000    0.036    0.000 fromnumeric.py:2046(clip)
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #42736
- [ ] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

```
from asv_bench.benchmarks.rolling import *

self = Engine()
self.setup("DataFrame", "float", sum, "cython", "sum")
self.time_rolling_methods("DataFrame", "float", sum, "cython", "sum")

%timeit self.time_rolling_methods("DataFrame", "float", sum, "cython", "sum")
150 µs ± 889 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <- PR
224 µs ± 3.18 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <- master
```

cc @mroeschke

### CODE REVIEW

File: pandas/core/window/rolling.py [hunk 1]
```diff
@@ -421,26 +420,40 @@ def _apply_blockwise(
...
+            obj = obj.T
 
-        if 0 != len(new_mgr.items) != len(mgr.items):
+        taker = []
+        res_values = []
+        for i, arr in enumerate(obj._iter_column_arrays()):
+            # GH#42736 operate column-wise instead of block-wise
+            try:
+                res = hfunc(arr)
+            except (TypeError, NotImplementedError):
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:437]: Will we allow these to raise once the nuisance column deprecation is enforced?

File: pandas/core/window/rolling.py [hunk 2]
```diff
@@ -421,26 +420,40 @@ def _apply_blockwise(
...
+                pass
+            else:
+                res_values.append(res)
+                taker.append(i)
+
+        df = type(obj)._from_arrays(
+            res_values,
+            index=obj.index,
+            columns=obj.columns.take(taker),
+            consolidate=False,
```
Comment (mroeschke) [inline pandas/core/window/rolling.py:447]: Can we not consolidate? I think every column should be float dtype

**Comment by mroeschke** [comment]
There's an `_apply_series` which looks like a 1D case of `_apply_blockwise`; can that be refactored into `_apply_blockwise`?

`_apply_blockwise` maybe should be renamed to `_apply_columnwise`

File: pandas/core/window/rolling.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/window/rolling.py:447]: we can, but its a pretty big performance hit (or at least it was before #43144; ill run some new timings.  best case would be to implement in 2D directly

File: pandas/core/window/rolling.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/window/rolling.py:437]: yes

**Comment by jbrockmendel** [comment]
Updated timings:

```
from asv_bench.benchmarks.rolling import *

self = Engine()
self.setup("DataFrame", "float", sum, "cython", "sum")
self.time_rolling_methods("DataFrame", "float", sum, "cython", "sum")

%timeit self.time_rolling_methods("DataFrame", "float", sum, "cython", "sum")
209 µs ± 8.64 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <- master
147 µs ± 3.32 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <- PR, consolidate=True
133 µs ± 1.95 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <- PR, consolidate=False
```

I'm OK with consolidating at that cost.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -2323,6 +2323,7 @@ def _from_arrays(
         index,
         dtype: Dtype | None = None,
         verify_integrity: bool = True,
+        consolidate: bool = True,
```
Comment (mroeschke) [inline pandas/core/frame.py:2326]: Is this needed now since it's always `True`?

Comment (jbrockmendel) [reply pandas/core/frame.py]: nope, removed

**Review by mroeschke — APPROVED** [review]
LGTM. Does this deserve a whatsnew note?

**Comment by jbrockmendel** [comment]
> Does this deserve a whatsnew note?

either way i guess, i can add something in the next "assorted follow-ups" PR

**Comment by mroeschke** [comment]
Sure that sounds good. Thanks @jbrockmendel

**Pull request merged**